pkg/pipeline/stream/destination_acker_test.go

```diff
@@ -16,61 +16,90 @@ package stream
 
 import (
 	"context"
+	"fmt"
+	"sync"
 	"testing"
 	"time"
 
 	"github.com/conduitio/conduit/pkg/connector/mock"
-	"github.com/conduitio/conduit/pkg/plugin"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
 	"github.com/conduitio/conduit/pkg/record"
 	"github.com/golang/mock/gomock"
 	"github.com/matryer/is"
 )
 
-func TestAckerNode_Run_StopAfterWait(t *testing.T) {
+func TestDestinationAckerNode_Cache(t *testing.T) {
 	is := is.New(t)
 	ctx := context.Background()
 	ctrl := gomock.NewController(t)
 	dest := mock.NewDestination(ctrl)
 
 	node := &DestinationAckerNode{
-		Name:        "acker-node",
+		Name:        "destination-acker-node",
 		Destination: dest,
 	}
 
+	in := make(chan *Message)
+	node.Sub(in)
+
 	nodeDone := make(chan struct{})
 	go func() {
 		defer close(nodeDone)
 		err := node.Run(ctx)
 		is.NoErr(err)
 	}()
 
-	// note that there should be no calls to the destination at all if we didn't
-	// receive any ExpectedAck call
+	const count = 1000
+	currentPosition := 0
+	dest.EXPECT().Ack(gomock.Any()).
+		DoAndReturn(func(ctx context.Context) (record.Position, error) {
+			pos := fmt.Sprintf("test-position-%d", currentPosition)
+			currentPosition++
+			return record.Position(pos), nil
+		}).Times(count)
+
+	var ackHandlerWg sync.WaitGroup
+	ackHandlerWg.Add(count)
+	for i := 0; i < count; i++ {
+		pos := fmt.Sprintf("test-position-%d", i)
+		msg := &Message{
+			Record: record.Record{Position: record.Position(pos)},
+		}
+		msg.RegisterAckHandler(func(msg *Message) error {
+			ackHandlerWg.Done()
+			return nil
+		})
+		in <- msg
+	}
 
-	// give the test 1 second to finish
-	waitCtx, cancel := context.WithTimeout(ctx, time.Second)
-	defer cancel()
-	node.Wait(waitCtx)
+	// note that there should be no calls to the destination at all if the node
+	// didn't receive any messages
+	close(in)
 
 	select {
-	case <-waitCtx.Done():
+	case <-time.After(time.Second):
 		is.Fail() // expected node to stop running
 	case <-nodeDone:
 		// all good
 	}
+
+	ackHandlerWg.Wait() // all ack handler should be called by now
```

Comment:
It is, that's how we would spot a failing test in this instance. I know, we could add a timeout, it is a hassle though and I feel it would needlessly complicate the test 😕  Maybe if we had those timeout utilities we wanted to add to the SDK, then we wouldn't have to reinvent the wheel every time we need a timeout for a waitgroup and this would be much simpler to add.

Reply:
Is there already an issue for adding the timeout utilities to the SDK? IMHO, we shouldn't allow the test to hang, but I also completely understand that it's a hassle.

Reply:
This is essentially the issue for adding the utilities to the SDK https://github.com/ConduitIO/conduit-connector-sdk/issues/26. Although even if we add them to the SDK, we would need to add them to Conduit separately (into the `foundation` package), since we shouldn't depend on utilities from the SDK in Conduit.